### ISSUE TITLE
Unsound usage of `TypeVar`s that are not in the class' `Generic` list in the `__init__` signature

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#2558]
**Describe the bug**
It's possible to use `TypeVar`s that are not part of the `Generic` list in the `__init__` signature of a class, effectively making the class generic over those unlisted `TypeVar`s, but having the resulting class unaware of that.

**To Reproduce**
```python3
# pyright: strict

from typing import Callable, TypeVar

T = TypeVar("T")
U = TypeVar("U")


class MyClass: # not generic
    def __init__(self, t: T, takes_t: Callable[[T], int]) -> None: # but actually generic over T
        self.t = t
        self.takes_t = takes_t

def takes_str(a: str) -> int:
    return 123

my_instance_str = MyClass(t="blah", takes_t=takes_str)
my_instance_str.takes_t(3.14) # calling takes_str with a float, but no error
```

**VS Code extension or command-line**
pyright extension version 1.1.186

**Comment by erictraut** [comment]
Hmm, I'm not sure what to do about this one.

Normally a TypeVar within a function scope must be solved within that function's scope. Any attempt to use the unsolved TypeVar outside the scope of that function results in an error. However, I've added special-case logic to allow unsolved function-scoped TypeVars within a Callable expression. This allows you to do things like:

```python
class MyClass:
    def set_identity_fn(self, identity: Callable[[T], T]) -> None:
        self.i_fn = identity

    def get_identity_fn(self) -> Callable[[T], T]:
        return self.i_fn
```

This is allowed because the Callable itself becomes the scope of the type variable, so it can be solved at a later time when the Callable is invoked.

I'm not sure how to close this hole without breaking these other use cases. I'll need to think about it more. Let me know if you have any suggestions.

**Comment by hmc-cs-mdrissi** [comment]
Not sure, but this feels like an issue of lack of quantifiers/scope.

def foo(fn: Callable[[T], T]) -> R:
  ...
  
type signature of this is Callable[[T], T] -> R, but is it forall T (Callable[[T], T] -> R) or is it, (for all T Callable[[T], T]) -> R. Passing a function that's [parametric polymorphic](https://en.wikipedia.org/wiki/Parametric_polymorphism) should be safe to leave unsolved for later. Passing a function that's not parametric but then treating it like it still has type vars later is unsafe.

I'm unsure of python solution without more explicit scoping for type variables. Where forall is intended to be is not clear in current signature. If I had to make a rough rule it'd be only parametric functions can be allowed to stay unsolved. If there was a ForAll type you could distinguish the two cases by having ForAll[Callable[[T], T]] means you want a parametric function and intend for (for all T Callable[[T], T]) -> R while normal Callable[[T], T] means you don't want that and intend for forall T (Callable[[T], T] -> R).

Without Forall annotation, for the current example my ideal expectation would be because takes_str is not parametric, when it's passed it is solved for and then my_instance_str.takes_t would have a type of (str) -> int instead of (T) -> int giving an error on last line, but not the line before it. Alternative option is mark previous line a type error with the reason being non-parametric function not allowed unless class itself is generic (as it effectively sets a type var on the class already).

I'm curious how many things actually need (for all T Callable[[T], T]) -> R. Feels more of a haskell thing than a python one. I'd be surprised if much code broke in either decision. I can remember only one time I've needed rank 2 type and just used type ignores giving up on it.

**Comment by Tomaz-Vieira** [comment]
I'm not sure I quite follow the identity function example; Clients of `get_identity_fn` would also be unable to resolve `T` if `MyClass` isn't explicitly generic over `T`, and that would cause the same issue from my example:

```python3
def int_to_int(a: int) -> int:
  return 123

my_intance.set_identity_fn(int_to_int) # pyright complains that set_identity_fn is unknown

str_identity: Callable[[str], str] = my_instance.get_identity_fn() # unsound behavior again
```
In fact, `set_identity_fn` isn't flagged upon definition, but it's usage is flagged with `type of set_identtiy_fn is unknown`

Maybe the problem is setting generic instance variables when the `TypeVar`s are not part of the class' `Generic` signature? Because that means that variable can be accessed later in some other context that can't possibly remember what `T` is supposed to be.

There is already the `reportUninitializedInstanceVariable`. I wonder if similar logic could be applied to report instance variables being initialized with generic types that are not covered by the class's `Generic` type vars.

**Comment by gjeusel** [comment]
Mmmh, not sure if it is linked, but I'm stumbling upon a similar problem when using a generic converter inside the `__init__`:


```python
from dataclasses import dataclass
from typing import Iterable, TypeVar, Union

T = TypeVar("T")
WannaBeList = Union[Iterable[T], T]


@dataclass
class TSMeta:
    tablename: str


def to_list(value: WannaBeList[T]) -> list[T]:
    if not hasattr(value, "__iter__"):
        return [value]  # type: ignore
    else:
        return list(value)  # type: ignore


class MyExc(Exception):
    def __init__(self, tsmetas: WannaBeList[TSMeta]):
        self.tsmetas = to_list(tsmetas)

    @property
    def tablenames(self) -> list[str]:
        # Cannot access member "tablename" for type Collection[TSMeta]
        return [tsmeta.tablename for tsmeta in self.tsmetas]  # <---- Give the above error


tsmetas = to_list(TSMeta(tablename="something"))
[tsmeta.tablename for tsmeta in tsmetas]  # works fine though
```

pyright 1.1.187

**Comment by erictraut** [comment]
@gjeusel, I don't think your issue is related. Your code is presenting the TypeVar constraint solver with an ambiguous case — a case where there are multiple ways to solve the constraints. It is picking one of the correct solutions, but you were expecting another one.

Consider this simplified example:

```python
def to_list(value: Iterable[T] | T) -> list[T]:
    ...

def func1(x: list[int] | int):
    y = to_list(x)
    reveal_type(y)
```

In this case, there are two possible solutions for the type of `y`. It can be `list[int]` or `list[list[int] | int]`. For various complex reason that I won't go into, pyright is choosing the second of these two solutions in this particular case. Choosing between TypeVar solutions in ambiguous cases is complex and involves some heuristics. Expressions that involve unions with a TypeVar present some of the trickiest situations for a type checker.

So I don't consider this a bug, and it's certainly not related to this issue. If you are not able to find a way to eliminate the ambiguity and would like me to look for ways to further tune the heuristics in this case, please file a separate issue. I can't make any guarantees because, well, heuristics are heuristics, and sometimes attempting to improve one use case will harm other use cases.

**Comment by gjeusel** [comment]
Woops, ok got it. Thanks for the reactivity, and sorry for the noise.

**Comment by hmc-cs-mdrissi** [comment]
> I'm not sure I quite follow the identity function example; Clients of `get_identity_fn` would also be unable to resolve `T` if `MyClass` isn't explicitly generic over `T`, and that would cause the same issue from my example:
> 
> ```python
> def int_to_int(a: int) -> int:
>   return 123
> 
> my_instance.set_identity_fn(int_to_int)
> 
> str_identity: Callable[[str], str] = my_instance.get_identity_fn() # unsound behavior again
> ```
> 
> Maybe the problem is setting generic instance variables when the `TypeVar`s are not part of the class' `Generic` signature? Because that means that variable can be accessed later in some other context that can't possibly remember what `T` is supposed to be.
> 
> There is already the `reportUninitializedInstanceVariable`. I wonder if similar logic could be applied to report instance variables being initialized with generic types that are not covered by the class's `Generic` type vars.

An example where set_identity_fn is type safe,

```python
def iden(x: T) -> T:
  print(x)
  return x

my_instance.set_identity_fn(iden)
str_identity: Callable[[str], str] = my_instance.get_identity_fn() 
```

It's completely type safe to cast to (str) -> str here because the identity fn given is [polymorphic](https://en.wikipedia.org/wiki/Parametric_polymorphism#Rank-1_(prenex)_polymorphism).

**Comment by erictraut** [comment]
Mypy appears to be inconsistent about this. It emits an error if a non-specialized callable is accessed through an instance variable but not when it's returned from a method or a property.

```python
from typing import Callable, TypeVar

T = TypeVar("T")

class MyClass:
    def __init__(self, identity: Callable[[T], T]) -> None:
        self.i_fn = identity

    def get_identity_fn(self) -> Callable[[T], T]:
        return self.i_fn

    @property
    def identity_fn(self) -> Callable[[T], T]:
        return self.i_fn


def foo(x: T) -> T:
    return x


m = MyClass(foo)

m.i_fn("hi")  # Error: Argument 1 has incomptible type "str"; expected "T"

m.identity_fn("hi")  # No Error

x = m.get_identity_fn()("hi")  # No Error
reveal_type(x)  # str
```

I agree that this presents a hole in the type system, but it's easy to fix in the code. It's ultimately related to the fact that TypeVar scoping in Python is so poorly defined. There has been some discussion of fixing this (e.g. adding dedicated syntax for specifying TypeVar scopes). Until that discussion comes to a conclusion, I think it's best to leave this as is for now.
